### PR TITLE
CRM-19150 Preserve contribution when associated membership is deleted

### PR DESCRIPTION
**Pull request by joseltorres**
Currently when a membership is deleted, if there is an associated contribution then the contribution also gets deleted. This adds the option to preserve the contrition. I suggest this be completed in two steps: 1) The ability the preserver the contribution through code (API and BAO) and 2) UI change to allow front-end users to check a box to preserve the associated contribution. This is step 1.

https://issues.civicrm.org/jira/browse/CRM-19150

### CODE REVIEW

**Comment by seamuslee001** [comment]
This looks good to merge :-)

**Comment by eileenmcnaughton** [comment]
Hmm - so if people want to preserve the contribution they need to re-write their code to add the preserve_contribution = 1

I'm find with us signalling a change on this & people having to change their code - as I think the default is dangerous.

But, if people have to change their code anyway then I'd prefer not to support this parameter on the api. Code that wants to do the double-down delete can just use chaining or separate delete calls. Having said that the fact there is a unit test on this swings me in favour of merging. :-)

Can you do a very short blog explaining the change & update https://wiki.civicrm.org/confluence/display/CRMDOC/API+changes

**Comment by joseltorres** [comment]
Maybe I wasn't clear, but the default behavior prior to this PR, was to delete the associated contribution. This PR will not change expected ('existing' might be more accurate) behavior for anyone's code, as not including this parameter in the API or BAO will have in the same result always had prior to this PR. What this PR does, is actually allow anyone using the membership delete API and BAO to be less data destructive if the choose to.

File: api/v3/Membership.php [hunk 1]
```diff
@@ -42,7 +59,17 @@
  *   API result array.
  */
 function civicrm_api3_membership_delete($params) {
-  return _civicrm_api3_basic_delete(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (isset($params['preserve_contribution'])) {
+    if (CRM_Member_BAO_Membership::del($params['id'], $params['preserve_contribution'])) {
+      return civicrm_api3_create_success(TRUE, $params);
```
Comment (eileenmcnaughton) [inline api/v3/Membership.php]: oh - these should be throwing exceptions - not create_error

**Pull request merged**